src/Tasks/SystemState.cs

```diff
@@ -547,13 +552,58 @@ out fileState.frameworkName
             frameworkName = fileState.frameworkName;
         }
 
-        /// <summary>
-        /// Cached implementation of GetDirectories.
-        /// </summary>
-        /// <param name="path"></param>
-        /// <param name="pattern"></param>
-        /// <returns></returns>
-        private string[] GetDirectories(string path, string pattern)
+        internal static SystemState DeserializePrecomputedCaches(string[] stateFiles, TaskLoggingHelper log, Type requiredReturnType, GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo)
+        {
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = true;
+
+            foreach (string stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sfBase = (SystemState)DeserializeCache(stateFile, log, requiredReturnType);
+                foreach (string s in sfBase.instanceLocalFileStateCache.Keys)
+                {
+                    if (!retVal.instanceLocalFileStateCache.ContainsKey(s))
+                    {
+                        FileState fileState = (FileState)sfBase.instanceLocalFileStateCache[s];
+                        // Verify that the assembly is correct
+                        if (File.Exists(s) && Assembly.LoadFile(s).ManifestModule.ModuleVersionId.Equals(fileState.ModuleVersionID))
+                        {
+                            // Correct file path and timestamp
+                            string fullPath = Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), stateFile));
```

Comment:
directory of stateFile, not current directory, right?

Reply:
Yes—directory of stateFile, and the relative path is s. (I should probably change that name, too.)